client.go

```diff
@@ -65,33 +65,21 @@ type Client struct {
 	ChallengeSolvers map[string]Solver
 }
 
-// ObtainCertificateUsingCSR obtains all resulting certificate chains using the given CSR, which
-// must be completely and properly filled out (particularly its DNSNames and Raw fields - this
-// usually involves creating a template CSR, then calling x509.CreateCertificateRequest, then
-// x509.ParseCertificateRequest on the output). The Subject CommonName is NOT considered.
-//
-// It implements every single part of the ACME flow described in RFC 8555 §7.1 with the exception
-// of "Create account" because this method signature does not have a way to return the updated
-// account object. The account's status MUST be "valid" in order to succeed.
-//
-// As far as SANs go, this method currently only supports DNSNames and IPAddresses on the csr.
-func (c *Client) ObtainCertificateUsingCSR(ctx context.Context, account acme.Account, csr *x509.CertificateRequest) ([]acme.Certificate, error) {
+type Finalizer interface {
+	Identifiers() []acme.Identifier
+	Finalize(context.Context) (*x509.CertificateRequest, error)
+}
+
+func (c *Client) ObtainCertificateUsingFinalizer(ctx context.Context, account acme.Account, finalizer Finalizer) ([]acme.Certificate, error) {
```

Comment:
One other option, I suppose, would be to pass in the identifiers and the finalizer.

What does your code that calls `ObtainCertificateUsingFinalizer()` look like? Maybe it does make more sense for you to just pass in the finalizer and put the IDs state in the finalizer. But then we have to have that extra `Identifiers()` method...

Reply:
I like the split responsibility you propose. The CSR is currently used for two purposes, which I wanted to decouple, but I did so in a way that's just slightly different, and still in a way that's coupled. I took the easy path 😉 Providing them as two separate arguments totally makes sense.

I don't have ready to share code publicly available now, but it's roughly like this:

```golang
func (c *Client) Sign(ctx context.Context, finalizer acmez.Finalizer) ([]*x509.Certificate, error) {
        ...
	acmeCerts, _ := c.client.ObtainCertificateUsingFinalizer(ctx, c.account, finalizer)
        ...
        return chain, nil
}
	
type myFinalizer struct {
	identifier string
	attester   attester.Attester
}

func (t *myFinalizer) Identifiers() []acmelib.Identifier {
	return []acmelib.Identifier{{Type: "permanent-identifier", Value: t.identifier}}
}

func (t *myFinalizer) Finalize(_ context.Context) (*x509.CertificateRequest, error) {
	return t.attester.CreateCertificateRequest(t.identifier, nil)
}

solver := solver.NewDeviceAttestationSolver(attester)
client, _ := myACME.New(ctx, solver)
finalizer := &myFinalizer{
    identifier: "ID",
    attester:   attester,
}
chain, _ := client.Sign(ctx, finalizer)
```

Providing the identifiers and the finalizer (source, injector, w/e) as two separate arguments is simple to implement on our side, so it sounds like a good approach to me.

Reply:
Awesome. That helps me a lot (to understand the context here).

Cool, so yeah... I think I am liking the idea of two different arguments as well.

Then we can reduce the interface to a single method. Maybe this?

```go
type CSRSource interface {
    CSR(context.Context) (*x509.CertificateRequest, error)
}
```

Reply:
Implemented in 40c9827bac02522581bf3b5c8df484535e1674fb.